### PR TITLE
feat(module:rate): emit hover change when leave

### PR DESCRIPTION
**Pull request by Nicoss54**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently when the rate is left no event emitter is sent. This prevent customisation like changing icon

Issue Number: #8415

## What is the new behavior?

hoverChange event is now sent when the user left the rate

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by OriginRing — APPROVED** [review]

**Comment by HyperLife1119** [comment]
I think it's a bug that OnHoverChange event is not triggered when the rate value is unchanged. In this case, we still need to trigger OnHoverChange event.
Let's fix it here :)
https://github.com/NG-ZORRO/ng-zorro-antd/blob/116346fdaaf1d5316b9836b3beda268538682e28/components/rate/rate.component.ts#L222-L232

**Comment by Nicoss54** [comment]
> I think it's a bug that OnHoverChange event is not triggered when the rate value is unchanged. In this case, we still need to trigger OnHoverChange event. Let's fix it here :)
> 
> https://github.com/NG-ZORRO/ng-zorro-antd/blob/116346fdaaf1d5316b9836b3beda268538682e28/components/rate/rate.component.ts#L222-L232

@HyperLife1119 it's done. Thanks again for the review

File: components/rate/rate.component.ts [hunk 1]
```diff
@@ -221,6 +221,7 @@ export class NzRateComponent implements OnInit, ControlValueAccessor, OnChanges
 
   onItemHover(index: number, isHalf: boolean): void {
     if (this.nzDisabled || (this.hoverValue === index + 1 && isHalf === this.hasHalf)) {
+      this.nzOnHoverChange.emit(this.hoverValue);
```
Comment (HyperLife1119) [inline components/rate/rate.component.ts:224]: If the state is disabled, the event should not be triggered.

Comment (Nicoss54) [reply components/rate/rate.component.ts]: done

File: components/rate/rate.component.ts [hunk 2]
```diff
@@ -220,10 +220,14 @@ export class NzRateComponent implements OnInit, ControlValueAccessor, OnChanges
...
   onItemHover(index: number, isHalf: boolean): void {
-    if (this.nzDisabled || (this.hoverValue === index + 1 && isHalf === this.hasHalf)) {
+    if (this.nzDisabled) {
       return;
     }
 
+    if (this.hoverValue === index + 1 && isHalf === this.hasHalf) {
+      return this.nzOnHoverChange.emit(this.hoverValue);
+    }
+
```
Comment (HyperLife1119) [inline components/rate/rate.component.ts:230]: I think the logic here would be this, WDYT :)

```ts
  onItemHover(index: number, isHalf: boolean): void {
    if (this.nzDisabled) {
      return;
    }

    if (this.hoverValue !== index + 1 || isHalf !== this.hasHalf) {
      this.hoverValue = index + 1;
      this.hasHalf = isHalf;
      this.updateStarStyle();
    }

    this.nzOnHoverChange.emit(this.hoverValue);
  }

```

Comment (Nicoss54) [reply components/rate/rate.component.ts:230]: Hi, I'm not against this proposal quite the contrary :), but if we do that we break the current behavior of the component isn't it ? Maybe we could do a little RFC ?

Comment (HyperLife1119) [reply components/rate/rate.component.ts:230]: https://stackblitz.com/edit/react-efsuqc?file=demo.tsx

Based on my observations of the behavior of the antd rate component, I think this is a bug fix.

If it is a breaking change, we will need the next major release before we can merge it.

Comment (Nicoss54) [reply components/rate/rate.component.ts]: @HyperLife1119 after investigation on my side you're right, so i made the change :). Thanks for your time and help

Comment (HyperLife1119) [reply components/rate/rate.component.ts]: Oh, I think you're missing some details:
- When disabled, events are not triggered.
- The event still needs to be triggered even if the rate value has not changed.

I think it's supposed to work like this:

```ts
  onItemHover(index: number, isHalf: boolean): void {
    if (this.nzDisabled) {
      return;
    }

    if (this.hoverValue !== index + 1 || isHalf !== this.hasHalf) {
      this.hoverValue = index + 1;
      this.hasHalf = isHalf;
      this.updateStarStyle();
    }

    this.nzOnHoverChange.emit(this.hoverValue);
  }
```

**Review by HyperLife1119 — APPROVED** [review]
LGTM.
@OriginRing WDYT?

File: components/rate/rate.component.ts (hunk 2 above)
Comment (HyperLife1119) [reply components/rate/rate.component.ts]: Great job, Nicoss :)

**Review by Laffery — APPROVED** [review]
LGTM

**Pull request merged**